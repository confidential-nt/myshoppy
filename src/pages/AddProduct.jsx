import React, { useEffect, useState } from "react";
import { useUserContext } from "../context/UserContext";
import { useNavigate } from "react-router-dom";

export default function AddProduct({ imageUploader, productRepository }) {
  const [loading, setLoading] = useState(false);
  const [complete, setComplete] = useState(false);
  const [file, setFile] = useState();
  const [name, setName] = useState("");
  const [price, setPrice] = useState("");
  const [category, setCategory] = useState("");
  const [desc, setDesc] = useState("");
  const [options, setOptions] = useState("");

  const { uid } = useUserContext();
  const navigate = useNavigate();

  useEffect(() => {
    if (!uid) {
      navigate("/");
    }
  }, [uid, navigate]);

  const handleSubmit = async (e) => {
    e.preventDefault();
    setLoading(true);
    await productRepository.insert({
      uid,
      imageURL: file,
      name,
      price,
      category,
      desc,
      options,
    });
    setLoading(false);
    setComplete(true);
    setFile(null);
    setTimeout(() => {
      setComplete(false);
    }, 3000);
  };

  const handleFileChange = (e) => {
    imageUploader.upload(e.target.files[0]).then(setFile);
  };

  return (
    <>
      {loading && (
        <div className="fixed top-0 left-0 w-screen h-screen bg-black/50 flex justify-center items-center">
          Uploading....
        </div>
      )}
      <div className="w-11/12 mr-auto ml-auto">
        <h2 className="text-center pt-2 pb-3 font-bold text-lg">
          새로운 제품 등록
        </h2>
        {complete && (
          <div className="flex justify-center">
            <span>✔ 성공적으로 제품이 추가되었습니다.</span>
          </div>
        )}
        {file && (
          <div className="w-full h-1/2 flex justify-center mb-3">
            <div className="w-64 h-96">
              <img
                className="w-full h-full object-cover"
                src={file}
                alt="등록할 제품"
              />
            </div>
          </div>
        )}
        <form className="flex flex-col" onSubmit={handleSubmit}>
          <label className="hidden" htmlFor="file-uploader">
            Upload file:
          </label>
          <input
            className="border pt-3 pb-3 pl-2 mb-2"
            type="file"
            accept="image/*"
            id="file-uploader"
            onChange={handleFileChange}
          />
          <input
            className="border pt-3 pb-3 pl-2 mb-2"
            type="text"
            placeholder="제품명"
            value={name}
            onChange={(e) => setName(e.target.value)}
          />
          <input
            className="border pt-3 pb-3 pl-2 mb-2"
            type="text"
            placeholder="가격"
            value={price}
            onChange={(e) => setPrice(e.target.value)}
          />
          <input
            className="border pt-3 pb-3 pl-2 mb-2"
            type="text"
            placeholder="카테고리"
            value={category}
            onChange={(e) => setCategory(e.target.value)}
          />
          <input
            className="border pt-3 pb-3 pl-2 mb-2"
            type="text"
            placeholder="제품 설명"
            value={desc}
            onChange={(e) => setDesc(e.target.value)}
          />
          <input
            className="border pt-3 pb-3 pl-2 mb-2"
            type="text"
            placeholder="옵션들(콤마(,)로 구분)"
            value={options}
            onChange={(e) => setOptions(e.target.value)}
          />
          <button
            className="bg-shoppypink text-white w-11/12 ml-auto mr-auto pt-2 pb-2 text-xl "
            type="submit"
          >
            제품 등록하기
          </button>
        </form>
      </div>
    </>
  );
}
