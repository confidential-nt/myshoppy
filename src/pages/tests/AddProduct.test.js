import { render, waitFor, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { withAllUserContexts, withRouter } from "../../tests/utils";
import { Route } from "react-router-dom";
import AddProduct from "../AddProduct";

describe("AddProduct", () => {
  const imageUploader = {
    upload: jest.fn(),
  };
  const productRepository = {
    insert: jest.fn(),
  };

  afterEach(() => {
    imageUploader.upload.mockReset();
    productRepository.insert.mockReset();
  });

  it("스냅샷 테스트", () => {
    const { asFragment } = renderAddProduct();
    expect(asFragment()).toMatchSnapshot();
  });

  it("제품 사진을 등록하면 제품 이미지가 나타나야한다", async () => {
    imageUploader.upload.mockImplementation(() => Promise.resolve("hello.png"));
    renderAddProduct();
    const file = new File(["hello"], "hello.png", { type: "image/png" });
    const input = screen.getByLabelText("Upload file:");
    await userEvent.upload(input, file);
    await waitFor(() => {
      expect(screen.getByRole("img")).toBeInTheDocument();
    });
  });

  it("제품을 등록하면 업로딩 중이라는 표시가 나타나야한다", async () => {
    productRepository.insert.mockImplementation(
      () =>
        new Promise((resolve, reject) => {
          setTimeout(() => resolve(null), 1000);
        })
    );

    renderAddProduct();
    const button = screen.getByRole("button", {
      name: "제품 등록하기",
    });
    await userEvent.click(button);

    await waitFor(() => {
      expect(screen.getByText("Uploading....")).toBeInTheDocument();
    });
  });

  it("제품 업로딩이 완료되면 완료되었다는 표시가 나타나야한다", async () => {
    productRepository.insert.mockImplementation(() => Promise.resolve(null));
    renderAddProduct();
    const button = screen.getByRole("button", {
      name: "제품 등록하기",
    });
    await userEvent.click(button);
    await waitFor(() => {
      expect(
        screen.getByText("✔ 성공적으로 제품이 추가되었습니다.")
      ).toBeInTheDocument();
    });
  });

  function renderAddProduct() {
    return render(
      withAllUserContexts(
        withRouter(
          <Route
            path="/"
            element={
              <AddProduct
                imageUploader={imageUploader}
                productRepository={productRepository}
              />
            }
          />
        ),
        null,
        {
          uid: "1",
          logUserIn: () => {},
        }
      )
    );
  }
});
