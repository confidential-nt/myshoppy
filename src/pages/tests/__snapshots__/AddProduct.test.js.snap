// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddProduct 스냅샷 테스트 1`] = `
<DocumentFragment>
  <div
    class="w-11/12 mr-auto ml-auto"
  >
    <h2
      class="text-center pt-2 pb-3 font-bold text-lg"
    >
      새로운 제품 등록
    </h2>
    <form
      class="flex flex-col"
    >
      <label
        class="hidden"
        for="file-uploader"
      >
        Upload file:
      </label>
      <input
        accept="image/*"
        class="border pt-3 pb-3 pl-2 mb-2"
        id="file-uploader"
        type="file"
      />
      <input
        class="border pt-3 pb-3 pl-2 mb-2"
        placeholder="제품명"
        type="text"
        value=""
      />
      <input
        class="border pt-3 pb-3 pl-2 mb-2"
        placeholder="가격"
        type="text"
        value=""
      />
      <input
        class="border pt-3 pb-3 pl-2 mb-2"
        placeholder="카테고리"
        type="text"
        value=""
      />
      <input
        class="border pt-3 pb-3 pl-2 mb-2"
        placeholder="제품 설명"
        type="text"
        value=""
      />
      <input
        class="border pt-3 pb-3 pl-2 mb-2"
        placeholder="옵션들(콤마(,)로 구분)"
        type="text"
        value=""
      />
      <button
        class="bg-shoppypink text-white w-11/12 ml-auto mr-auto pt-2 pb-2 text-xl "
        type="submit"
      >
        제품 등록하기
      </button>
    </form>
  </div>
</DocumentFragment>
`;
